tools/spxls/internal/server/diagnostic.go

```diff
@@ -0,0 +1,170 @@
+	gopmodload "github.com/goplus/mod/modload"
+)
+
+// See https://microsoft.github.io/language-server-protocol/specifications/lsp/3.18/specification#textDocument_diagnostic
+func (s *Server) textDocumentDiagnostic(params *DocumentDiagnosticParams) (*DocumentDiagnosticReport, error) {
+	diags, err := s.diagnose()
+	if err != nil {
+		return nil, err
+	}
+
+	return &DocumentDiagnosticReport{Value: &RelatedFullDocumentDiagnosticReport{
+		FullDocumentDiagnosticReport: FullDocumentDiagnosticReport{
+			Kind:  string(DiagnosticFull),
+			Items: diags[params.TextDocument.URI],
+		},
+	}}, nil
+}
+
+// See https://microsoft.github.io/language-server-protocol/specifications/lsp/3.18/specification#workspace_diagnostic
+func (s *Server) workspaceDiagnostic(params *WorkspaceDiagnosticParams) (*WorkspaceDiagnosticReport, error) {
+	diags, err := s.diagnose()
+	if err != nil {
+		return nil, err
+	}
+
+	var items []WorkspaceDocumentDiagnosticReport
+	for file, fileDiags := range diags {
+		items = append(items, Or_WorkspaceDocumentDiagnosticReport{
+			Value: &WorkspaceFullDocumentDiagnosticReport{
+				URI: DocumentURI(file),
+				FullDocumentDiagnosticReport: FullDocumentDiagnosticReport{
+					Kind:  string(DiagnosticFull),
+					Items: fileDiags,
+				},
+			},
+		})
+	}
+	return &WorkspaceDiagnosticReport{Items: items}, nil
+}
+
+// diagnose performs diagnostic checks for spx source files and returns diagnostics for each file.
+func (s *Server) diagnose() (map[DocumentURI][]Diagnostic, error) {
+	spxFiles, err := s.spxFiles()
+	if err != nil {
+		return nil, fmt.Errorf("get spx files failed: %w", err)
+	}
+
+	diags := make(map[DocumentURI][]Diagnostic, len(spxFiles))
+
+	fset := goptoken.NewFileSet()
+	gpfs := vfs.NewGopParserFS(s.workspaceRootFS)
+	mainPkgFiles := make(map[string]*gopast.File)
+	for _, file := range spxFiles {
+		documentURI := s.toDocumentURI(file)
+		diags[documentURI] = nil
+
+		f, err := gopparser.ParseFSEntry(fset, gpfs, file, nil, gopparser.Config{
+			Mode: gopparser.AllErrors | gopparser.ParseComments,
+		})
+		if err != nil {
+			// Handle parse errors.
+			var parseErr gopscanner.ErrorList
+			if errors.As(err, &parseErr) {
+				for _, e := range parseErr {
+					diags[documentURI] = append(diags[documentURI], Diagnostic{
+						Range: Range{
+							Start: FromGopTokenPosition(e.Pos),
+							End:   FromGopTokenPosition(e.Pos),
+						},
+						Message: e.Msg,
+					})
+				}
+				continue
+			}
+
+			// Handle code generation errors.
+			var codeErr *gogen.CodeError
+			if errors.As(err, &codeErr) {
+				position := codeErr.Fset.Position(codeErr.Pos)
+				diags[documentURI] = append(diags[documentURI], Diagnostic{
+					Range: Range{
+						Start: FromGopTokenPosition(position),
+						End:   FromGopTokenPosition(position),
+					},
+					Message: codeErr.Error(),
+				})
+				continue
+			}
+
+			// Handle unknown errors.
+			diags[documentURI] = append(diags[documentURI], Diagnostic{
+				Severity: SeverityError,
```

Comment:
上边的 parse error & generation error 需要指定 `Severity` 吗？

Reply:
哦上面没写可能是因为当时看到 `Severity` 默认 `SeverityError`：

```ts
/**
 * The diagnostic's severity. To avoid interpretation mismatches when a
 * server is used with different clients it is highly recommended that
 * servers always provide a severity value. If omitted, it’s recommended
 * for the client to interpret it as an Error severity.
 */
severity?: DiagnosticSeverity;
```

但写到这个位置时可能忘了这个设定😅

我来都加上吧，我看 spec 里也这么建议

Reply:
OK

Reply:
done https://github.com/goplus/builder/compare/484dd28fb724ad66ef2cc57d2d811221125bdb15..d947f3fbb0775b3dbaa09b3437a26ae74299dc59